core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -37,25 +40,35 @@
 
     private final DiscoveryNode node;
 
-    private final List<ShardRouting> shards;
+    private final Map<ShardId, ShardRouting> shards;
 
     public RoutingNode(String nodeId, DiscoveryNode node) {
-        this(nodeId, node, new ArrayList<ShardRouting>());
+        this(nodeId, node, new LinkedHashMap<>()); // LinkedHashMap to preserve order
     }
 
-    public RoutingNode(String nodeId, DiscoveryNode node, List<ShardRouting> shards) {
+    public RoutingNode(String nodeId, DiscoveryNode node, Map<ShardId, ShardRouting> shards) {
         this.nodeId = nodeId;
         this.node = node;
         this.shards = shards;
     }
 
+    // public for tests
+    public RoutingNode(String nodeId, DiscoveryNode node, Iterable<ShardRouting> shardsCollection) {
+        this(nodeId, node);
+        for (ShardRouting shardRouting : shardsCollection) {
+            ShardRouting previousValue = shards.put(shardRouting.shardId(), shardRouting);
+            assert previousValue == null : "Cannot have two different shards with same shard id on same node";
+        }
+
+    }
```

Comment:
ok, I've instead generalized the earlier constructor by adding a varargs ShardRouting argument.

I've also changed the signature of the constructor used by RoutingNodes to make explicit that we need a LinkedHashMap preserving iteration order (and made that one package visible).